snuba/query/processors/type_converters/__init__.py

```diff
@@ -0,0 +1,114 @@
+        self.columns = columns
+        column_match = Or([String(col) for col in columns])
+
+        literal = Param("literal", LiteralMatch(AnyMatch(str)))
+
+        operator = Param(
+            "operator",
+            Or(
+                [
+                    String(op)
+                    for op in FUNCTION_TO_OPERATOR
+                    if op not in (ConditionFunctions.IN, ConditionFunctions.NOT_IN)
+                ]
+            ),
+        )
+
+        in_operators = Param(
+            "operator",
+            Or((String(ConditionFunctions.IN), String(ConditionFunctions.NOT_IN))),
+        )
+
+        col = Param("col", ColumnMatch(None, column_match))
+
+        self.__condition_matcher = Or(
+            [
+                FunctionCallMatch(operator, (literal, col)),
+                FunctionCallMatch(operator, (col, literal)),
+            ]
+        )
+
+        self.__in_condition_matcher = FunctionCallMatch(
+            in_operators,
+            (col, Param("tuple", FunctionCallMatch(String("tuple"), None)),),
+        )
+
+    def strip_column_alias(self, exp: Expression) -> Expression:
+        assert isinstance(exp, Column)
+        return Column(
+            alias=None, table_name=exp.table_name, column_name=exp.column_name
+        )
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        query.transform_expressions(
+            self.process_expressions, skip_transform_condition=True
+        )
+
+        condition = query.get_condition()
+        if condition:
+            query.set_ast_condition(condition.transform(self.process_condition))
+
+    def process_condition(self, exp: Expression) -> Expression:
+        match = self.__condition_matcher.match(exp)
+        if match:
+            return FunctionCall(
+                exp.alias,
+                match.string("operator"),
+                (
+                    self.strip_column_alias(match.expression("col")),
+                    self.translate_literal(match.expression("literal")),
+                ),
+            )
+
+        in_condition_match = self.__in_condition_matcher.match(exp)
+
+        if in_condition_match:
+            tuple_func = in_condition_match.expression("tuple")
+            assert isinstance(tuple_func, FunctionCall)
+            new_tuple_func = FunctionCall(
+                tuple_func.alias,
+                tuple_func.function_name,
+                parameters=tuple(
+                    [self.translate_literal(lit) for lit in tuple_func.parameters]
+                ),
+            )
+            return FunctionCall(
+                exp.alias,
+                in_condition_match.string("operator"),
+                (
+                    self.strip_column_alias(in_condition_match.expression("col")),
+                    new_tuple_func,
+                ),
+            )
+
+        return exp
+
+    @abstractmethod
+    def translate_literal(self, exp: Expression) -> Expression:
```

Comment:
Since this is implemented by the subclasses, but it is only the parent class that can guarantee the parameter is always a literal (as the pattern is in the parent class), I would move the assertion in the parent class and make `translate_literal` take a `Literal` instead of an Expression